CIP-0021/CIP-0021.md

```diff
@@ -0,0 +1,115 @@
+
+**Certificates**
+
+Certificates of type `genesis_key_delegation` and `move_instantaneous_rewards_cert` are not supported and must not be included.
+
+If a transaction contains a pool registration certificate, then it must not contain:
+
+- any other certificate;
+- any withdrawal;
+- mint entry.
+
+It is allowed to arbitrarily combine other supported certificate types, but all the certificates included in a transaction must have the same type of stake credential i.e. either 0 - `addr_keyhash` or 1 - `scripthash`. The stake credential type must also be consistent with the type used for withdrawals.
+
+**Withdrawals**
+
+Since withdrawals are represented as a map of reward accounts, withdrawals also need to be sorted in accordance with the specified canonical CBOR format. A transaction must not contain duplicate withdrawals. All withdrawals included in a transaction must have the same type of stake credential i.e. either 0 - `addr_keyhash` or 1 - `scripthash`. The stake credential type must also be consistent with the type used for certificates.
+
+### Transaction witnesses
+
+There are two limits on the number of witnesses:
+
+- an absolute limit of `UINT16_MAX`, i.e. 65535;
+- a relative limit dependent on the transaction body (essentially one witness per each input, each withdrawal and each certificate in a typical transaction).
+
+## Reasoning
+
+### Canonical CBOR serialization format
+
+As HW wallets don't return the whole serialized transaction, a common CBOR serialization is needed so that software wallets and other tools interacting with HW wallets are be able to deterministically reproduce the transaction body built and signed by the HW wallet.
+
+The specified canonical CBOR format is consistent with how certain other data are serialized (e.g. Plutus script data in Alonzo) and allows the use of standard CBOR libraries out of the box.
+
+### Transaction body
+
+**Multiassets**
+
+Allowing duplicate `policy_id`s (or `asset_name`s) might lead to inconsistencies between what is displayed to the user and what ends up being included in the transaction. All policies (or asset names) would be shown to the user, but trying to include duplicate policies in the transaction body might result in either the first entry or the second entry being included in the transaction.
+
+**Certificates**
+
+Combining withdrawals and pool registration certificates isn't allowed because both are signed by staking keys by pool owners. If it was allowed to have both in a transaction then the witnesses provided by pool owners might inadvertently serve as a witness for a withdrawal for the owner's account.
+
+**Withdrawals**
+
+Similarly to multiassets, allowing duplicate withdrawals might lead to inconsistencies between what is displayed to the user and what ends up being included in the transaction.
```

Comment:
Well, both will be included in the tx (otherwise tx body hash will not fit), but it is unclear how nodes and other tools interpret cbor maps with duplicate entries, and there is no point in allowing them anyway.

Reply:
I've udpated it - https://github.com/vacuumlabs/CIPs/pull/3/commits/e3e04d9115ed34ac2767446b16b9fd7dd4599a9c#diff-03ac633fb24ddb2e5f1ab3ab8360277e59e4aef208f07856922c702ad436f094R101